### PR TITLE
application: serial_lte_modem: Automation of LwM2M requests

### PR DESCRIPTION
**Pull request by junqingzou**
Add `CONFIG_SLM_CARRIER_AUTO_CONTROL` to control automatical handling of below requests from LwM2M in SLM, instead of by host MCU:
 - Link down/up
 - Power off modem
 - Reboot

By default this CONFIG is enabled.

JIRA-ticket: NCSDK-28267
JIRA-ticket: NCSDK-28431

### CODE REVIEW

File: applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c
Comment (junqingzou) [reply applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c:231]: SBM asked for a delay of 5s and I offered 100ms as I'm not sure about context of callback in LwM2M stack. SLM_UART_RESPONSE_DELAY is 50ms.

**Review by lucasvdp — COMMENTED** [review]
This solution only works for a succesfull fota. I think we should add a setting, either compile time or in the form of an AT command. This setting could allow SLM to automatically do everything or let everything up to the user.
It will be hard to capture all the scenarios after a FOTA_START event because some events require us to do something after the FOTA_SUCCES or ERROR FOTA_FAIL event.
Also for other Carrier we get these events that require an action during bootstrap. It would make sense to also allow these to be automated.

**Comment by junqingzou** [comment]
> This solution only works for a succesfull fota. I think we should add a setting, either compile time or in the form of an AT command. This setting could allow SLM to automatically do everything or let everything up to the user. It will be hard to capture all the scenarios after a FOTA_START event because some events require us to do something after the FOTA_SUCCES or ERROR FOTA_FAIL event. Also for other Carrier we get these events that require an action during bootstrap. It would make sense to also allow these to be automated.

CONFIG_SLM_CARRIER_AUTO_REBOOT is added to control the behavior of auto reboot after APP FOTA download. Other scenario can refer to this config but not in the scope of this PR, I think.

File: applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c
Comment (tomi-font) [inline applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c:207]:
```suggestion
```

File: applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c
Comment (tomi-font) [reply applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c:238]: `SLM_UART_RESPONSE_DELAY` could be increased to 100ms if you think 100ms is needed. I was thinking it'd be nice to use it here because it's for the same purpose than other places using it.

**Review by tomi-font — APPROVED** [review]

File: applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c
Comment (junqingzou) [reply applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c:238]: @tomi-font SBM just accepted current behavior without further requesting time (5min) for host side to prepare for a reboot. So I'll silently change the delay from 100ms to 50ms in next submit.

**Comment by junqingzou** [comment]
> `CONFIG_SLM_CARRIER_AUTO_REBOOT` is a good improvement, but I think we should also get rid of the `fota_started` variable. It is still not unset when a fota does not end with a reboot (app fota finished) or link up (modem fota finished). We can simply allow all reboots, link_up, link_down and power_off. Maybe rename `CONFIG_SLM_CARRIER_AUTO_REBOOT` to `CONFIG_SLM_CARRIER_AUTO_HANDLE_EVENTS`? I understand wanting to limit the scope of this PR to reboot only, but I don't like having a half working inbetween state and my suggestion is a quicker fix than keeping proper track of the `fota_started`.

Let's get this in first for SBM IOT. For further improvement or design change, I'll leave it to Kacper.

File: applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c
Comment (tomi-font) [inline applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c:156]:
```suggestion
#if defined(CONFIG_SLM_CARRIER_FOTA_AUTOMATION)
```
Same for the other instances of this.

File: applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c
Comment (tomi-font) [inline applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c:252]: Why not rather put this [here](https://github.com/nrfconnect/sdk-nrf/pull/16360/files#diff-1c276f4a29001e10696ebbf7251866a78b4593747545d5a151ca4c4e44040dd4R214)?

File: applications/serial_lte_modem/src/lwm2m_carrier/Kconfig
Comment (tomi-font) [inline applications/serial_lte_modem/src/lwm2m_carrier/Kconfig:35]: `during a FOTA`?

**Review by tomi-font — CHANGES_REQUESTED** [review]

File: applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c
Comment (junqingzou) [reply applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c:252]: No, after URC is sent

File: applications/serial_lte_modem/src/lwm2m_carrier/Kconfig
Comment (junqingzou) [reply applications/serial_lte_modem/src/lwm2m_carrier/Kconfig:35]: Automation for FOTA context only

Comment (tomi-font) [reply applications/serial_lte_modem/src/lwm2m_carrier/Kconfig:35]: Yeah, that's why I think it would be good to make this more clear in the Kconfig option help message.

**Comment by kacperradoszewski** [comment]
This works for the simple success/failure FOTA cases but does not behave well with some other scenarios, that don't always apply to SBM but do affect other carriers, which can lead to the `fota_started` variable to be out of sync (RAT switch, some errors that happen before we push anything to DFU, and so on). So similarly to @lucasvdp, I also suggest removing that one and handle the requests unconditionally, perhaps under a common `CONFIG_SLM_CARRIER_AUTO_REQUEST_HANDLING`, `CONFIG_SLM_CARRIER_AUTO_HANDLE_EVENTS` or something like that.

LINK_UP and LINK_DOWN events are only really sent during FOTA and bootstrap process (to write PSKs into the modem, this does not apply to SBM), so I don't think it hurts to automate these altogether, since we already wanted to automate the bootstrap too. It also won't hurt to automate REBOOT.

I don't mind if this is controlled with a KConfig or an AT command, I think it will suffice if this is a static setting.

**Comment by junqingzou** [comment]
> This works for the simple success/failure FOTA cases but does not behave well with some other scenarios, that don't always apply to SBM but do affect other carriers, which can lead to the `fota_started` variable to be out of sync (RAT switch, some errors that happen before we push anything to DFU, and so on). So similarly to @lucasvdp, I also suggest removing that one and handle the requests unconditionally, perhaps under a common `CONFIG_SLM_CARRIER_AUTO_REQUEST_HANDLING`, `CONFIG_SLM_CARRIER_AUTO_HANDLE_EVENTS` or something like that.
> 
> LINK_UP and LINK_DOWN events are only really sent during FOTA and bootstrap process (to write PSKs into the modem, this does not apply to SBM), so I don't think it hurts to automate these altogether, since we already wanted to automate the bootstrap too. It also won't hurt to automate REBOOT.
> 
> I don't mind if this is controlled with a KConfig or an AT command, I think it will suffice if this is a static setting.

What you said is automate LINK_UP/LINK_DOWN/REBOOT with an static CONFIG?
SLM has unconditional POWER OFF/SHUTDOWN MODEM/INIT MODEM as of now

**Comment by kacperradoszewski** [comment]
> What you said is to automate LINK_UP/LINK_DOWN/REBOOT with an static CONFIG beyond FOTA scenario?

Yes, this will simplify the code and ensure we don’t enter some weird out-of-sync states.

> SLM has unconditional POWER OFF/SHUTDOWN MODEM/INIT MODEM as of now

Modem shutdown and init are unconditional because there are no AT commands to handle those I think? (I might be mistaken, perhaps SLM has some mechanism)
Don’t remember the reason for different handling of power off vs link down/up, but we can potentially put that one under that config too.

**Comment by junqingzou** [comment]
> > What you said is to automate LINK_UP/LINK_DOWN/REBOOT with an static CONFIG beyond FOTA scenario?
> 
> Yes, this will simplify the code and ensure we don’t enter some weird out-of-sync states.
> 
> > SLM has unconditional POWER OFF/SHUTDOWN MODEM/INIT MODEM as of now
> 
> Modem shutdown and init are unconditional because there are no AT commands to handle those I think? (I might be mistaken, perhaps SLM has some mechanism) Don’t remember the reason for different handling of power off vs link down/up, but we can potentially put that one under that config too.

OK, I'll make the automation for REBOOT/LINK UP/LINK DOWN unconditional.
Modem shutdown and init were made unconditional when rebootless MFW DFU was added in LwM2M by Stig as I remember.

File: applications/serial_lte_modem/src/lwm2m_carrier/Kconfig
Comment (tomi-font) [inline applications/serial_lte_modem/src/lwm2m_carrier/Kconfig:29]:
```suggestion
	bool "Automatic handling of LwM2M requests"
```

File: applications/serial_lte_modem/src/lwm2m_carrier/Kconfig
Comment (tomi-font) [inline applications/serial_lte_modem/src/lwm2m_carrier/Kconfig:33]:
```suggestion
	  Enable automatic power off, reboot, link up or link down when requested by the LwM2M stack.
```

File: applications/serial_lte_modem/src/lwm2m_carrier/Kconfig
Comment (tomi-font) [inline applications/serial_lte_modem/src/lwm2m_carrier/Kconfig:34]:
```suggestion
	  If this is disabled, the host MCU must perform these operations via AT commands when requested.
```

File: applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c
Comment (tomi-font) [inline applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c:160]:
```suggestion
		err = slm_util_at_printf("AT+CFUN=4");
```
or maybe even (and same for the other occurences)
```suggestion
		err = slm_util_at_printf("AT+CFUN=4");
		if (err) {
			err = -1;
		}
```

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (tomi-font) [inline applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:97]:
```suggestion
  Request to perform an application reboot, i.e. ``AT#XRESET``.
```

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (tomi-font) [inline applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:50]:
```suggestion
  Request to set the modem to minimum functional mode, i.e. ``AT+CFUN=0``.
```

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (tomi-font) [inline applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:42]:
```suggestion
  Request to set the modem to flight functional mode, i.e. ``AT+CFUN=4``.
```

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (tomi-font) [inline applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:34]:
```suggestion
  Request to set the modem to full functional mode, i.e. ``AT+CFUN=1``.
```

**Review by tomi-font — CHANGES_REQUESTED** [review]

**Comment by kacperradoszewski** [comment]
I imagine the

**Comment by kacperradoszewski** [comment]
Sorry about that, I've mis-clicked... I imagine the SLM tests will need to be adjusted now that SLM_CARRIER_AUTO_CONTROL is enabled by default @tomi-font @MarkusLassila I can help if needed

**Comment by junqingzou** [comment]
> Sorry about that, I've mis-clicked... I imagine the SLM tests will need to be adjusted now that SLM_CARRIER_AUTO_CONTROL is enabled by default @tomi-font @MarkusLassila I can help if needed

If not wanted I could enable it in NIDD overlay instead. But POWEROFF will still behave differently

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (tomi-font) [inline applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:97]:
```suggestion
  Request to perform an application reboot, i.e. ``AT#XRESET``.
```

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (tomi-font) [inline applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:37]: Uppercase R is missing on every line that you modified.

**Review by tomi-font — CHANGES_REQUESTED** [review]
Getting there!

**Comment by kacperradoszewski** [comment]
> > Sorry about that, I've mis-clicked... I imagine the SLM tests will need to be adjusted now that SLM_CARRIER_AUTO_CONTROL is enabled by default @tomi-font @MarkusLassila I can help if needed
> 
> If not wanted I could enable it in NIDD overlay instead. But POWEROFF will still behave differently

I think with the different POWEROFF behaviour, the tests will probably pass but might lead to a reboot without a modem power off first (potentially problematic over several reboots).

**Review by kacperradoszewski — APPROVED** [review]

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (junqingzou) [reply applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:97]: `e.g.` means there are other ways than `AT#XRESET`

**Comment by junqingzou** [comment]
This PR have helped in passing SB FOTA over LwM2M/NIDD (1)APP split-DFU (2)MFW DFU and (3) MFW DFU Abort test cases.

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (tomi-font) [reply applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:97]: Like what?

Comment (junqingzou) [reply applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:97]: .AT#XCARRIER="reboot"
.nRESET system reset (pin reset)
.ENABLE toggle
.nrfjprog -r (soft reset)
.powercycle

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (tomi-font) [reply applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:37]: This is not resolved.

Comment (junqingzou) [reply applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:37]: Ah, just locally, next commit after CI is updated

**Review by MarkusLassila — APPROVED** [review]
Looks good.

Created PR for moving the LwM2M carrier tests to nightly. This can be merged when that get's in.

Created a separate PR for adapting the changes to CI and ran this against it. To be merged after this has gone in.

Edit: @junqingzou, LwM2M carrier tests moved to nightly. This should now pass the CI.

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (wiba-nordic) [inline applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:34]: We don't use "i.e." and "e.g." in the documentation, so this will need to be opened up.
```suggestion
  Request to set the modem to full functional mode, equivalent to the ``AT+CFUN=1` AT command`.
```

Should be applied to all the lines where these are used.

**Review by wiba-nordic — COMMENTED** [review]
Does this need a changelog entry?

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (junqingzou) [reply applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:34]: Not equivalent but must issue the AT command as mentioned.
Change to another wording?

Comment (wiba-nordic) [reply applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:34]: Ah, so it doesn't automatically do it. Then, would this be correct:
```suggestion
  Request to set the modem to full functional mode, such as using the ``AT+CFUN=1`` AT command.
```
Alternatively, if the AT command is basically the only option, we can change the order:
```suggestion
  Request to use the ``AT+CFUN=1`` AT command to set the modem to full functional mode.
```

File: applications/serial_lte_modem/doc/CARRIER_AT_commands.rst
Comment (tomi-font) [inline applications/serial_lte_modem/doc/CARRIER_AT_commands.rst:102]: Still missing uppercase letters here.

**Review by tomi-font — CHANGES_REQUESTED** [review]

**Review by tomi-font — APPROVED** [review]

**Review by wiba-nordic — APPROVED** [review]

**Pull request merged**
